File: clients/python/bigdb/results.py
# Copyright 2026 Bany
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""What an answer is, once it has been read.

Every class here is frozen and carries `raw`, the decoded body it came from. The `raw` is not
decoration: it is what lets a field a newer server adds be reachable without a release of this
client, which matters for a surface that is still growing.

Kinds and engine names are the server's spellings, kept verbatim and never translated. There
are two vocabularies - `?kind=signed` on the way in, `"signedint"` on the way back
(`routes/mod.rs::parse_kind` against `json.rs`'s `Debug`-lowercased render) - and a client that
normalised them into one would be inventing a third that neither side speaks.
"""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass, field
from typing import Any

__all__ = [
    "Count",
    "Extreme",
    "FieldInfo",
    "Group",
    "Groups",
    "PqlAnswer",
    "ProjectionRow",
    "ProjectionRows",
    "Ready",
    "RecordPage",
    "Schema",
    "SqlResult",
    "Sum",
    "TableInfo",
    "TextResult",
    "TupleGroup",
    "Tuples",
    "WriteResult",
]

_EMPTY: Mapping[str, Any] = {}


# --------------------------------------------------------------------------------------------
# Schema.
# --------------------------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class FieldInfo:
    """One field, as `/schema` describes it."""

    name: str
    #: The server's spelling: one of `set mutex bool int decimal timequantum signedint float32
    #: float64 date datetime`. Note this is *not* the vocabulary `create_field` takes.
    kind: str
    bit_depth: int
    #: Present only for a decimal. A decimal without its scale is an integer wearing a different
    #: name - `price > 5` means `> 500` on a field with two of them.
    scale: int | None = None
    #: Present only for a time quantum, as single characters: `Y`, `M`, `D`, `H`.
    granularity: tuple[str, ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class TableInfo:
    """One table and its fields."""

    name: str
    #: `bitmap`, `bitmap+columnar` or `columnar`, verbatim.
    engine: str
    fields: tuple[FieldInfo, ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)

    def field_named(self, name: str) -> FieldInfo | None:
        """The field by that name, or `None`."""
        return next((f for f in self.fields if f.name == name), None)


@dataclass(frozen=True, slots=True)
class Schema:
    """Every table this node knows about."""

    tables: tuple[TableInfo, ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)

    def table(self, name: str) -> TableInfo | None:
        """The first table with that name.

        **First, not the only one.** `TableInfo` carries a `database` on the server
        (`crates/big-embed/src/schema.rs`) that `json::schema` does not render, so two
        same-named tables in two databases arrive here indistinguishable. Across a single
        database this is a lookup; across several it is a guess, and that is the server's to
        fix rather than this client's to paper over.
        """
        return next((t for t in self.tables if t.name == name), None)


# --------------------------------------------------------------------------------------------
# `/sql`.
# --------------------------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class SqlResult:
    """Columns and rows, the shape a SQL client is written to read."""

    columns: tuple[str, ...] = ()
    rows: tuple[tuple[Any, ...], ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)

    def dicts(self) -> list[dict[str, Any]]:
        """The rows as dictionaries, for the reading that does not care about order."""
        # strict=False: a row the server sent shorter than its own header is its
        # business to explain, and losing the other rows over it helps nobody.
        return [dict(zip(self.columns, row, strict=False)) for row in self.rows]

    def scalar(self) -> Any:
        """The one cell, for a statement that answers with one.

        Raises `ValueError` rather than returning `None` on an empty result, because `None` is
        also a perfectly good cell value and the two would be indistinguishable.
        """
        if len(self.rows) != 1 or len(self.rows[0]) != 1:
            raise ValueError(
                f"this result is {len(self.rows)} rows by {len(self.columns)} columns, not one cell"
            )
        return self.rows[0][0]


@dataclass(frozen=True, slots=True)
class TextResult:
    """A result set the statement asked for in `FORMAT TSV|TSVWithNames|CSV|CSVWithNames`.

    Handed back as text rather than parsed. Reading it back into rows would mean a CSV reader
    in this client disagreeing with the writer in `crates/big-sql/src/shape.rs` about quoting -
    and a caller who asked for CSV asked for it because something downstream reads CSV.
    """

    content_type: str
    text: str


# --------------------------------------------------------------------------------------------
# `/table/{t}/query` - one class per shape `json::value_paged` can write.
# --------------------------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class Count:
    """`{"count": n}`."""

    value: int
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class Sum:
    """`{"sum": n}`.

    Signed, unsigned and real sums all arrive in this one shape on purpose: JSON numbers are
    signed, so a client reading it needs to know nothing about how the field was declared.
    """

    value: int | float
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class Extreme:
    """`{"value": n | null}` - a min or a max.

    `None` is nothing matched, which is a different answer from a total of nothing. That is why
    the server writes `null` here rather than `0`.
    """

    value: int | float | None
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class RecordPage:
    """`{"records": [...], "next": n | null}`.

    `next` is set only when the page came back exactly as long as the limit asked for, so it
    means "there may be more after this id" rather than "there is more".
    """

    records: tuple[int, ...] = ()
    next: int | None = None
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class TupleGroup:
    """One entry of `{"tuples": [...]}`: a list of keys, and the value for that combination."""

    keys: tuple[str | None, ...]
    value: PqlAnswer
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class Tuples:
    """`{"tuples": [...]}` - a pair (or wider) grouping asked for in the query language."""

    tuples: tuple[TupleGroup, ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class ProjectionRow:
    """One entry of `{"rows": [...]}`: a record and its projected cells."""

    record: int
    values: tuple[Any, ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class ProjectionRows:
    """`{"rows": [...]}` - a projection asked for in PQL rather than in SQL.

    One object per record rather than an array of cells, because this route's answers name what
    they hold; the result-set shape belongs to `/sql`, where the client asked for a table.
    """

    rows: tuple[ProjectionRow, ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class Group:
    """One entry of `{"groups": [...]}`.

    `key` is the bucket's own name and `row` the number it is addressed by. A keyed group
    carries the string beside the row id because a row id is meaningless without the dictionary
    that issued it; a calendar bucket has no dictionary and writes itself out as the date it
    stands for.
    """

    key: str | None
    row: int
    value: PqlAnswer
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


@dataclass(frozen=True, slots=True)
class Groups:
    """`{"groups": [...]}`."""

    groups: tuple[Group, ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)


#: What `Client.query` can answer with. A union with no discriminant on the wire - the shapes
#: are told apart only by which key is present - so `decode.pql` checks in a fixed order and
#: refuses a shape it does not recognise rather than handing back a dict pretending to be typed.
PqlAnswer = Count | Sum | Extreme | RecordPage | Tuples | ProjectionRows | Groups


# --------------------------------------------------------------------------------------------
# Writes and probes.
# --------------------------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class WriteResult:
    """What an import or a delete managed."""

    written: int
    #: The copies that did not take it, described the way the server describes them -
    #: `"a-spare (0..1) (connection refused)"`. **Strings, not record ids**: an unreachable
    #: replica does not fail a write, it is named here and marked behind, and `POST /repair`
    #: catches it up. A non-empty `missed` means the write is on fewer copies than it should be.
    missed: tuple[str, ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)

    @property
    def complete(self) -> bool:
        """Whether every copy took it."""
        return not self.missed


@dataclass(frozen=True, slots=True)
class Ready:
    """`GET /ready`, which is never authenticated.

    The cluster fields - `serving`, `term`, `leader`, `behind` - are present only on a node that
    is in one. `wire` is data here and never a gate: this client does not send `x-big-wire`,
    which `routes/mod.rs::mismatched` checks on `/internal/*` alone.
    """

    status: str = ""
    tables: int = 0
    txn_id: int = 0
    pages: int = 0
    node: str = ""
    shards: str = ""
    version: str = ""
    wire: int = 0
    serving: bool | None = None
    term: int | None = None
    leader: str | None = None
    behind: tuple[str, ...] = ()
    raw: Mapping[str, Any] = field(default_factory=lambda: _EMPTY, repr=False)

File: clients/go/results.go
// Copyright 2026 Bany
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package bigdb

import "encoding/json"

// Every shape the server answers with, typed.
//
// # Why nothing here is map[string]any
//
// encoding/json decodes a number into any as a float64, and record ids are u64. Run it and see:
//
//	in : {"records":[18446744073709551615],"next":9007199254740993}
//	out: {"next":9007199254740992,"records":[18446744073709552000]}
//
// The largest record id comes back wrong by 1615 and the cursor by one. So every field that can
// hold a record id, a count or a sum is declared with the width it has, and the escape hatch
// for a field this build has not heard of is Raw - a json.RawMessage, which is lossless,
// allocates nothing, and lets a caller decode it into a type of their own.

// Answer is what POST /table/{t}/query returns. It is one of the concrete types in this file.
//
// The wire form is a union with no discriminant - seven shapes told apart only by which key is
// present (big_http::json::value_paged) - so this is a sealed interface and decoding walks the
// keys in a fixed order. A shape this build has never heard of comes back as a *ProtocolError
// rather than as something pretending to be typed.
type Answer interface{ isAnswer() }

// CountAnswer is {"count": n}.
type CountAnswer struct {
	Count uint64
	Raw   json.RawMessage
}

// SumAnswer is {"sum": n}.
//
// Signed, unsigned and real sums all arrive under the same key, because JSON numbers are signed
// and a client reading {"sum":...} should not have to know how the field was declared. Value
// holds an integer sum; Real holds it when the field is a float, and IsReal says which.
type SumAnswer struct {
	Value  int64
	Real   float64
	IsReal bool
	Raw    json.RawMessage
}

// ValueAnswer is {"value": n} or {"value": null}.
//
// Null is "nothing matched", which is a different answer from a total of nothing - so the
// absence is a bool rather than a zero.
type ValueAnswer struct {
	Value   int64
	Real    float64
	IsReal  bool
	Present bool
	Raw     json.RawMessage
}

// RecordsAnswer is {"records": [...], "next": id|null}.
type RecordsAnswer struct {
	Records []uint64
	// Next is the id to send back as `after`, or nil when this page is the last.
	Next *uint64
	Raw  json.RawMessage
}

// TupleItem is one pair grouping: the keys that name it and the value at it.
type TupleItem struct {
	Keys  []GroupKey
	Value Answer
}

// TuplesAnswer is {"tuples": [...]}.
type TuplesAnswer struct {
	Tuples []TupleItem
	Raw    json.RawMessage
}

// GroupKey names one bucket. Key is the string a keyed group carries beside its row id, or the
// date a calendar bucket stands for; Row is the number it is addressed by.
//
// A bucket names itself: a row id is meaningless without the dictionary that issued it, so the
// server hands back both. See big_http::json::named_at.
type GroupKey struct {
	Key    string
	HasKey bool
	Row    int64
}

// GroupItem is one group and its value.
type GroupItem struct {
	GroupKey
	Value Answer
}

// GroupsAnswer is {"groups": [...]}.
type GroupsAnswer struct {
	Groups []GroupItem
	Raw    json.RawMessage
}

// ProjectionRow is one record and the cells projected from it.
type ProjectionRow struct {
	Record uint64
	Values []json.RawMessage
}

// RowsAnswer is {"rows": [...]} - a projection asked for in the query language.
//
// The cells stay as RawMessage: a projection cell is one of five things (absent, an integer, a
// real, a string, a list of strings) and which one is the field's business, not this decoder's.
type RowsAnswer struct {
	Rows []ProjectionRow
	Raw  json.RawMessage
}

func (*CountAnswer) isAnswer()   {}
func (*SumAnswer) isAnswer()     {}
func (*ValueAnswer) isAnswer()   {}
func (*RecordsAnswer) isAnswer() {}
func (*TuplesAnswer) isAnswer()  {}
func (*GroupsAnswer) isAnswer()  {}
func (*RowsAnswer) isAnswer()    {}

// SQLResult is what POST /sql returns.
//
// # Two shapes, and the client never guesses which
//
// A statement's FORMAT clause changes both the body and the Content-Type
// (big_sql::shape). So Columns and Rows are filled for application/json, and Text and
// ContentType for everything else. The decision is made from the response's Content-Type and
// never by scanning the statement - a client that parsed SQL to find a FORMAT clause would be a
// second parser to keep in step with big_sql.
type SQLResult struct {
	Columns []string
	Rows    [][]json.RawMessage

	// Text is the body when the format is not JSON, with ContentType saying which it is.
	Text        []byte
	ContentType string

	Raw json.RawMessage
}

// IsText reports whether this answer came back in a format other than JSON.
func (r *SQLResult) IsText() bool { return r.Columns == nil && r.Text != nil }

// WriteResult is what an import or a delete returns.
//
// Missed carries the keys the server could not place. It is absent from the body when empty,
// which is why it is a slice and not a count: the names are the useful part.
type WriteResult struct {
	Count  uint64
	Missed []string
	Raw    json.RawMessage
}

// RecordPage is one page of GET /table/{t}/records.
//
// Next is the cursor. Note the server's own caveat (big_http::json::records): a full page
// always reports a cursor, even when it happens to be the last one, because the listing asked
// for exactly `limit` ids and cannot see whether a further one exists without another read.
// A scan therefore ends with one empty page, not with a nil Next on the last full one.
type RecordPage struct {
	Records []uint64
	Next    *uint64
	Raw     json.RawMessage
}

// FieldInfo is one field in the schema.
//
// Kind is the server's spelling, kept as it arrived and never translated. Be aware that the
// read and write vocabularies differ: /schema renders "signedint" but the create-field route's
// parse_kind takes "signed", and posting ?kind=signedint is a 400. This client passes Kind
// through in both directions rather than pretending there is one vocabulary.
//
// Scale is present only for a decimal - a decimal without its scale is an integer wearing a
// different name - and Granularity only for a time quantum.
type FieldInfo struct {
	Name        string
	Kind        string
	BitDepth    uint32
	Scale       int64
	HasScale    bool
	Granularity []string
	Raw         json.RawMessage
}

// TableInfo is one table in the schema.
type TableInfo struct {
	Name   string
	Engine string
	Fields []FieldInfo
	Raw    json.RawMessage
}

// Field finds a field by name, or nil.
func (t *TableInfo) Field(name string) *FieldInfo {
	for i := range t.Fields {
		if t.Fields[i].Name == name {
			return &t.Fields[i]
		}
	}
	return nil
}

// Schema is the answer to GET /schema.
type Schema struct {
	Tables []TableInfo
	Raw    json.RawMessage
}

// Table finds a table by name, or nil.
//
// # Why this is not a trustworthy lookup across databases
//
// big_db::TableInfo carries the database a table belongs to, but big_http::json::schema does
// not render it. So two tables with the same name in two databases arrive here identical, and
// this returns whichever came first. A caller who needs to tell them apart has to ask the
// server with ?database= set rather than sort it out on this side.
func (s *Schema) Table(name string) *TableInfo {
	for i := range s.Tables {
		if s.Tables[i].Name == name {
			return &s.Tables[i]
		}
	}
	return nil
}

// Ready is the answer to GET /ready.
//
// Serving is the part a probe can act on: a node that has lost touch with the agreement refuses
// requests for its range, and a balancer that keeps sending them is sending them somewhere that
// will answer 503. It is still ready - the engine is fine and the node is one promotion away -
// so it is a field rather than a failure, and it is absent on a single node.
//
// Version and Wire are two different questions during a rolling upgrade: Version is which build
// this is, Wire is what it speaks to other nodes. Two nodes with different Version and the same
// Wire run side by side; two with different Wire do not.
type Ready struct {
	Status  string
	Tables  int
	TxnID   uint64
	Pages   uint64
	Node    string
	Shards  string
	Version string
	Wire    int

	Serving    bool
	HasServing bool
	Term       uint64
	Leader     string
	Behind     []string

	Raw json.RawMessage
}
